### PR TITLE
Update TTKG firmware version from status message

### LINKED ISSUE #4836: Update Gateway attributes for TTKG version
#### Summary

We should update the version information of The Things Kickstarter Gateways in the Identity Server when it changes.

#### What is already there? What do you see now?

We update `station`, `firmware` and `package` attributes for Basic Station gateways, so that we can show gateway owners what firmware versions their gateways are running.

#### What is missing? What do you want to see?

We don't track the firmware versions of Kickstarter gateways.

#### How do you propose to implement this?

When receiving a status message with version info, compare it to the gateway attributes, and update them if changed.

#### How do you propose to test this?

Just plug in a TTKG.

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

Closes #4836 

#### Changes

- Update `model` and `firmware` attributes from the TTKG platform field.
- Add some delay/jitter to prevent sudden updates.
- Make sure to update only on change.
- Tests

#### Testing

UT

##### Regressions

Don't expect any.

#### Notes for Reviewers

1. The one edge case here is when we connect the gateway in `unregistered` mode. The `Update` call would then fail. We could just print the error (as this PR does). The other option is to add a `registered` flag to the connection and update only if set. Let me know which one you prefer.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/gatewayserver/config.go [hunk 1]
```diff
@@ -57,6 +57,8 @@ type Config struct {
 	UpdateGatewayLocationDebounceTime time.Duration `name:"update-gateway-location-debounce-time" description:"Debounce time for gateway location updates from status messages"`
 	UpdateConnectionStatsDebounceTime time.Duration `name:"update-connection-stats-debounce-time" description:"Time before repeated refresh of the gateway connection stats"`
 
+	UpdateVersionInfoDelay time.Duration `name:"update-version-info-delay" description:"Maximum time to wait to update version information. A Jitter of 100% is applied for randomization"`
```
Comment (adriansmares) [inline pkg/gatewayserver/config.go]: 100% ? I'd suggest 15-25%.

File: pkg/gatewayserver/gatewayserver.go [hunk 2]
```diff
@@ -802,6 +819,7 @@ func (gs *GatewayServer) handleUpstream(ctx context.Context, conn connectionEntr
 		case msg := <-conn.Status():
 			ctx = events.ContextWithCorrelationID(ctx, fmt.Sprintf("gs:status:%s", events.NewCorrelationID()))
 			val = msg
+
```
Comment (adriansmares) [inline pkg/gatewayserver/gatewayserver.go]: Remove.

File: pkg/gatewayserver/gatewayserver.go [hunk 3]
```diff
@@ -972,6 +990,36 @@ func (gs *GatewayServer) handleLocationUpdates(ctx context.Context, conn connect
 	}
 }
 
+// handleVersionInfoUpdates updates gateway attributes with version info.
+// This function runs exactly once; only for the first status message of each connection, since version information does (should) not change within the same connection.
```
Comment (adriansmares) [inline pkg/gatewayserver/gatewayserver.go]: Just leave it as `should`.

File: pkg/gatewayserver/gatewayserver.go [hunk 4]
```diff
@@ -1012,6 +1060,7 @@ func (gs *GatewayServer) UnclaimDownlink(ctx context.Context, ids ttnpb.GatewayI
 	return gs.UnclaimIDs(ctx, &ids)
 }
 
+// ValidateGatewayID implements IO.
```
Comment (adriansmares) [inline pkg/gatewayserver/gatewayserver.go]: `io.Server`.

File: pkg/gatewayserver/gatewayserver.go [hunk 5]
```diff
@@ -972,6 +990,36 @@ func (gs *GatewayServer) handleLocationUpdates(ctx context.Context, conn connect
...
 
+// handleVersionInfoUpdates updates gateway attributes with version info.
+// This function runs exactly once; only for the first status message of each connection, since version information does (should) not change within the same connection.
+func (gs *GatewayServer) handleVersionInfoUpdates(ctx context.Context, conn connectionEntry) {
+	select {
+	case <-ctx.Done():
+		return
+	case <-conn.VersionInfoChanged():
+		status, _, ok := conn.StatusStats()
+		if ok && status.Versions["model"] != "" && status.Versions["firmware"] != "" {
```
Comment (adriansmares) [inline pkg/gatewayserver/gatewayserver.go]: Add alias for `status.Versions` and `conn.Gateway().Attributes`.

If `conn.Gateway().Attributes` is `nil`, this panics.
Also make sure that `status.Versions` is not `nil` (I'm not sure if we initialize this everywhere).

File: pkg/gatewayserver/gatewayserver.go [hunk 6]
```diff
@@ -972,6 +990,36 @@ func (gs *GatewayServer) handleLocationUpdates(ctx context.Context, conn connect
...
+				"model":    status.Versions["model"],
+				"firmware": status.Versions["firmware"],
+			}
+			d := random.Jitter(gs.config.UpdateVersionInfoDelay, 1)
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(d):
+			}
+			err := gs.entityRegistry.UpdateAttributes(conn.Context(), *conn.Gateway().Ids, *&conn.Gateway().Attributes, attributes)
```
Comment (adriansmares) [inline pkg/gatewayserver/gatewayserver.go]: `*&` does not copy the object. Make a clone explicitly before the call, or inside `UpdateAttributes`.

**Comment by adriansmares** [comment]
Also wouldn't this be shorter and easier if you compare the versions in `updateConnStats` ? Then you don't need a whole separate flow just for the first status. 

Regarding unregistered: Printing the error suffices.

File: pkg/gatewayserver/gatewayserver.go (hunk 5 above)
Comment (KrishnaIyer) [reply pkg/gatewayserver/gatewayserver.go]: I don't think so. As far as I'm aware, you can read from an uninitialised map.
https://go.dev/play/p/TGaXMKqyTXc

**Comment by KrishnaIyer** [comment]
> Also wouldn't this be shorter and easier if you compare the versions in updateConnStats ? Then you don't need a whole separate flow just for the first status.

Yes the reason is that these two update different registries, different debouncing, different life time (only once v/s multiple). I think this is cleaner than adding more if/else blocks in the `updateConnStats`.

File: pkg/gatewayserver/config.go (hunk 1 above)
Comment (KrishnaIyer) [reply pkg/gatewayserver/config.go]: The default delay is 5s. So -100 is 0 and +100 is 10s. That's gives us a 10s range. Using 25%, the range is 3.75 - 6.25. This means that this call will always wait for at least 3.75s and I don't want that.

File: pkg/gatewayserver/gatewayserver.go (hunk 5 above)
Comment (adriansmares) [reply pkg/gatewayserver/gatewayserver.go]: Correct, only write panics on `nil` maps.

Also turn the `if` around and return if it holds

```
if !ok || versions["model"] == "" || versions["firmware"] == "" {
    return
}
```

File: pkg/gatewayserver/config.go (hunk 1 above)
Comment (adriansmares) [reply pkg/gatewayserver/config.go]: The average user will wait 5 seconds anyway. The fact that the lower bound is 'lower' does not change the expected value. But ok, I'm fine with a `[0s, 10s]` range.

**Review by adriansmares — APPROVED** [review]
